default/AI/ResearchAI.py

```diff
@@ -453,28 +492,45 @@ def generate_research_orders():
     rng = random.Random()
     rng.seed(fo.getEmpire().name + fo.getGalaxySetupData().seed)
 
-    research_reqs = calculate_research_requirements(empire)
+    calculate_research_requirements(empire)
     total_rp = empire.resourceProduction(fo.resourceType.research)
 
     if total_rp <= 0: # No RP available - no research.
         return
 
+    raw_priorities = {}
     priorities = {}
+    on_path_to = {}
     for tech_name in fo.techs():
-        priority = get_priority(rng, tech_name)
+        this_tech = fo.getTech(tech_name)
+        if not this_tech:
+            continue
+        raw_priorities[tech_name] = get_priority(rng, tech_name)
+    for tech_name, priority in raw_priorities.iteritems():
         if not tech_is_complete(tech_name) and priority >= 0:
-            turn_needed = float(research_reqs[tech_name][1]) / total_rp
-            priorities[tech_name] = float(priority) / turn_needed / max(1, (turn_needed - 12) / 4 + 1)**2
+            turn_needed = max(research_reqs[tech_name][2], math.ceil(float(research_reqs[tech_name][1]) / total_rp))
+            adjusted_priority = float(priority) / turn_needed / max(1, (turn_needed - 12) / 4 + 1)**2
```

Comment:
> Then again, after these techs are researched, if I started with the slow techs, I will be able to research the other ones much faster (assuming that our RP are growing in the meantime). The slow but cheap techs will still take maximum time.

Hmm, I actually think we might be able to resolve this with math, or at least make substantial progress on it.  I'll foreshadow that the math, at least when applied to the range of tech characteristics I think most common in FO,  seems to *not* favor my position.  (So surely there must be some error in the math, please point it out for me. :smiley: )

Consider 3 techs, all with the same value payoff, all with the same total cost.  One of them can be completed in N turns by consuming all the RP we have; the other two can be completed in time 2N in parallel.  So at the end of 3N turns we have them all finished (we can consider your increasing-empire-RP point afterwards).  At that point we will either have had one tech's benefit for 2N turns, or two techs' benefit for N turns.  If we consider that any such tech benefit is like some investment property whose cumulative contribution to empire capabilities goes up each turn with a slow exponential (taking k turns to double), then our balance point is where 2^(2N/k) = 2 * 2^(N/k), or, by my math, where N = k

*So if N is less than our doubling time constant k, which I think is very likely the case for the majority of our techs, then we should do the two techs in parallel first.*  Now, I did sort of slip an extra factual premise into the middle of the above math, namely that the techs don't just all have the same total cost and initial benefit value, but also that the rate of appreciation on the benefit is the same.   These key characteristics-- initial benefit value, and rate of appreciation, all get tossed into the same priority pot with our current approach, and of course we would be hard pressed to actually assign particular numbers for these things even if we adopted a framework for it.  I suppose that is why the 'original' Research AI, at the time I inherited it, simply researched cheapest first, no other consideration, and managed to somewhat muddle through things like that.  Of course, that approach as also implicitly assuming that everything had the same benefit, and the AI did much better with some guided research.

***I guess the key takeaway is that, as a general matter, we should indeed assign little consideration, perhaps even none, directly to total projected completion time, and instead focus on minimizing per-turn costs (and better base priorities).***  That first part is easy enough, and then we can focus on making the base priority assessments be more accurate.   It's very nice to have a framework that can do some level of automatic adaptation to tech characteristics, and I think our handling for ship related tech is a really great example of that, but I think we need a tidier way of contributing our current "expert knowledge" about existing techs.  We do it some already here, but I'd like to make it easier/cleaner, and have an idea for that I will try to implement.